### PR TITLE
Add bridge API to export all wallet private keys

### LINKED ISSUE #755: Exporting key lists does not work
When trying to export key lists for a bridge/LMDB wallet, I hit several problems in sequence:

1. **Backup** — Clicking Backup did not open the backup dialog.
2. **Other options** — After fixing that, clicking Other options did nothing.
3. **Export key lists** — After fixing that, clicking Export key lists did not open the key list dialog.
4. **Private keys in the dialog** — After fixing that, the key list dialog opened but private keys (PrivBase58 / PrivHexBE) did not appear for bridge wallet addresses. In some cases the dialog crashed with NameErrors (e.g. `PRIVKEYBYTE` or `computeChecksum` undefined in AddressUtils).

So exporting key lists does not work for bridge/LMDB wallets without additional fixes.

**Environment:** Debian Linux (kernel 6.1.0-41-amd64), built from source following the project’s Linux build instructions (build_instructions/linux/linux.md) with system deps (capnproto, lmdb) and built deps (libbtc, libwebsockets).

### PR DESCRIPTION
**Pull request by 151henry151**
This description has been revised to reflect changes to the PR since the initial submission (rebased onto `0.97_rc2`, scoped to bridge + test only, and reworked per review feedback). The GUI changes are in #770.

Add a bridge API to export every private key in a wallet in one batch call, with a test that exercises encrypted export. This is the bridge-only first step toward fixing the export key list for bridge wallets; the GUI follow-up is a separate PR.

The export walks all address and asset accounts and returns every computed private key under a single `lockDecryptedContainer(lbd)` lock, using the post-rc2 unlock-lifetime API. The unlock callback id is passed as a plain `Types.CallbackId` with no dedicated request struct. `BridgeWalletTests.ExportPrivateKeysEncrypted` asserts the full key set is returned, rejects a wrong passphrase, and restores the wallet to compare the re-export against the original export (the restore step surfaces the known uncomputed-key edge case goatpig offered to fix alongside merge). This PR also fixes unrelated PySide6 signal connections in `DlgBackupCenter` so backup buttons work under PySide6.

Contributes to #755.

Tested by rebuilding `CppBridge` and `gtest/BridgeTests`, running `BridgeWalletTests.ExportPrivateKeysEncrypted` (export-all passes; the restore full-count assertion fails as expected pending the upstream edge-case fix), and manually exercising the full key-export GUI flow on the stacked GUI branch with an encrypted wallet in offline mode.

### CODE REVIEW

**Comment by 151henry151** [comment]
@goatpig I reworked PR #756 around a batch `exportPrivateKeys` bridge call that sets a `BridgePassphrasePrompt` unlock lambda with a callback id, decrypts eligible private keys under a single `lockDecryptedContainer()` lifetime, and caches the result in the key-list dialog.

During manual GUI testing I also noticed PySide6 prints `_pythonToCppCopy: Cannot copy-convert ... QMouseEvent to C++` on some clickable-label interactions, but it did not block the export flow.

**Comment by goatpig** [comment]
Hey,

Thanks for the PR. Could you please rebase it on top of #768 before review?

**Comment by 151henry151** [comment]
> Thanks for the PR. Could you please rebase it on top of #768 before review?

Rebased on #768 -- should be good to go. I didn't repeat the manual GUI testing yet.

**Comment by goatpig** [comment]
Hey, sorry about the delay with the review.

While you have indeed rebased your commit on top of `0.97_rc2` branch, your PR is still against the `dev` branch. Could you please base the PR itself on top of the rc2 PR (#768)?

**Comment by 151henry151** [comment]
> While you have indeed rebased your commit on top of `0.97_rc2` branch, your PR is still against the `dev` branch. Could you please base the PR itself on top of the rc2 PR (#768)?

Done.

File: armoryengine/PyBtcWallet.py [hunk 1]
```diff
@@ -142,6 +143,13 @@ def __init__(self, *, wltId=None, accId=None, proto=None):
       elif proto != None:
          self.loadFromProto(proto)
 
+      # Bridge wallets: C++ manages decryption via passphrase lambda.
+      # Legacy wallets: start locked when encryption is enabled.
```
Comment (goatpig) [inline armoryengine/PyBtcWallet.py]: Legacy wallets are not loaded anymore

File: armoryengine/PyBtcWallet.py [hunk 2]
```diff
@@ -683,6 +695,12 @@ def getLinearAddrList(self, withImported=True, withAddrPool=False):
                addrList.append(addr)
       return addrList
 
+   def hasAnyImported(self):
+      for addr in self.addrMap.values():
+         if addr.chainIndex == -2:
```
Comment (goatpig) [inline armoryengine/PyBtcWallet.py]: This is legacy wallet logic, it doesn't apply anymore. I'll add a call do to perform that check to this branch shortly.

File: cppForSwig/capnp/Bridge.capnp [hunk 3]
```diff
@@ -489,6 +493,7 @@ struct WalletRequest {
 
       getUnlockTime                 @19: Void;
       forkWatchingOnly              @20: Types.CallbackId;
+      exportPrivateKeys             @21: ExportPrivateKeysRequest;
```
Comment (goatpig) [inline cppForSwig/capnp/Bridge.capnp]: The argument is a plain callback id, there is no need for dedicated struct

File: cppForSwig/gtest/BridgeTests.cpp [hunk 4]
```diff
@@ -1125,6 +1125,146 @@ namespace {
...
+         auto fromBridge = reader.getRoot<Codec::Bridge::FromBridge>();
+
+         //could be a reply (success or error) or a notification
+         if (fromBridge.which() == Codec::Bridge::FromBridge::REPLY) {
+            auto reply = fromBridge.getReply();
+            out.success = reply.getSuccess();
+            if (!out.success) {
+               out.error = reply.getError();
+            } else {
+               auto walletReply = reply.getWallet();
```
Comment (goatpig) [inline cppForSwig/gtest/BridgeTests.cpp:1211]: This expects to parse a valid reply in 2 different places.

The test is deterministic, it either unlocks the wallet successfully and a successful reply shows up before the cleanup notif, or it fails to unlock and it's cleanup -> success == false.

The fetch flow should reflec this distinction.

File: cppForSwig/gtest/BridgeTests.cpp [hunk 5]
```diff
@@ -5499,6 +5639,56 @@ TEST_F(BridgeWalletTests, ChangeWalletPassphrase)
...
+
+   auto wallets = loadWallets(bridge_);
+   ASSERT_EQ(wallets.size(), 1ULL);
+   ASSERT_EQ(wallets.begin()->first, walletId);
+
+   /* 3. export with correct passphrase — S2 */
+   {
+      auto result = exportPrivateKeys(bridge_, walletId, privPass, true);
+      ASSERT_TRUE(result.success) << "export failed: " << result.error;
+      ASSERT_GE(result.keys.size(), 1ULL);
```
Comment (goatpig) [inline cppForSwig/gtest/BridgeTests.cpp]: The wallet was generated with 4 addresses but the test expects only 1 key to be exported. As mentioned in the CppBridge comment, it should export all private keys indiscriminately.

It should also restore the wallet, export and check vs the original export list. This may lead to an edge case where the wallet has not computed certain private keys. It is fine to deliver a test that fails to run so long as it covers this use case. If that happens, I'll fix the edge case and merge the PR along with the fix.

File: qtdialogs/DlgWalletDetails.py [hunk 6]
```diff
@@ -537,7 +537,8 @@ def execRemoveDlg(self):
          pass  # not sure that I don't handle everything in the dialog itself
 
    def execKeyList(self):
-      if self.wlt.useEncryption and self.wlt.isLocked:
+      isBridge = bool(getattr(self.wlt, 'bridgeWalletObj', None))
```
Comment (goatpig) [inline qtdialogs/DlgWalletDetails.py]: I don't understand what this is for. ArmoryQt does not load its own wallets anymore, it's essentially a dumb wrapper around CppBridge, in order to render a pyqt GUI. All wallets live in the bridge, and whether a wallet has private key encryption should not matter, the same call should be used, with a callbackId.

You do not need to `exec_` explicitly on DlgUnlockWallet. If CppBridge needs to prompt the user for a passphrase, it will spawn the unlock dialog via the callback flow. So long as there is a valid instance of DlgUnlockWallet tied to the callbackId fed to CppBridge, it will spawn that dialog and cleanup it up according to the notification flow.

You can look at tx signing (armoryengine/Transcation.py:130) or backup creation (qtdialogs/DlgBackupCenter.py:400) for examples.

File: ui/WalletFrames.py [hunk 7]
```diff
@@ -1115,7 +1115,8 @@ def clickedDoIt(self):
          isBackupCreated = self.main.makeWalletCopy(
             self, self.wlt, 'Encrypt', 'encrypt')
       elif self.optIndivKeyListTop.isChecked():
-         if self.wlt.useEncryption and self.wlt.isLocked:
+         isBridge = bool(getattr(self.wlt, 'bridgeWalletObj', None))
```
Comment (goatpig) [inline ui/WalletFrames.py]: Same comment as with DlgWalletDetails

File: qtdialogs/DlgShowKeyList.py [hunk 8]
```diff
@@ -10,41 +10,78 @@
...
-      for addr in self.wlt.getLinearAddrList(withAddrPool=True):
-         self.addrCopies.append(addr.copy())
-      self.rootKeyCopy = self.wlt.addrMap['ROOT'].copy()
-
-      backupVersion = BACKUP_TYPE_135A
-      testChain = DeriveChaincodeFromRootKey(self.rootKeyCopy.binPrivKey32_Plain)
-      self.needChaincode = (not testChain == self.rootKeyCopy.chaincode)
-      if not self.needChaincode:
-         backupVersion = BACKUP_TYPE_135C
+      if self.isBridgeWallet:
```
Comment (goatpig) [inline qtdialogs/DlgShowKeyList.py]: There are only bridge wallets

File: qtdialogs/DlgShowKeyList.py [hunk 9]
```diff
@@ -10,41 +10,78 @@
...
+         # Bridge wallets: private keys are fetched async; start locked.
+         self.havePriv = False
+         self.rootKeyCopy = None
+         self.needChaincode = False
```
Comment (goatpig) [inline qtdialogs/DlgShowKeyList.py]: 1.35a wallets carry a root chaincode.

Look at comments for `enum class BackupType : int` in Backups.h for more details.

**Comment by goatpig** [comment]
I've reviewed the code. In general it works but it has 2 major issues:

1. You are being misled by the legacy code.

Armory up to 0.96.5 had a python implementation for wallets, and a SWIG wrapper around a shared lib for DB interfacing. For 0.97, all wallet code has been redone in a cpp binary (CppBridge) that also deals with the DB.

As a general rule, I do not touch python code files that lack the relevant CppBridge feature. This is leading to 2 confusions on your end: 
a. you are trying to maintain support for "legacy wallet" as in legacy code.
b. this is masking the difference between legacy wallets (Armory 1.35a/c address chains/backups) and new wallets (Armory 2.00 chains/backups).

There is no legacy code operating anymore, but the GUI has to support the differences between Armory 1.35 and 2.00 chains when it comes to backups and key lists.

2. The export code should support rendering every address and/or private key in the wallet. This is a critical feature for exports, as users may be manipulating the keys from an offline instance, which cannot inform them on address use state.

This will likely be a complex effort, as currently the python side does NOT load all assets in the wallet, only those in use. Either PyBtcWallet will need a bunch of modifications, or the export will have to grab both public and private keys as well as relevant metadata and ignore the PyBtcWallet object when rendering keys.

Maybe it would be better to make this 2 PRs. So long as this PR can demonstrate fetching all private keys in the wallet in the bridge test, I'm happy to merge that as a first step.

-------------

Lastly, the passphrase lambda lifetime has been annoying me for a while so I pushed a commit that ties its lifetime to the decrypted data lock. It will impact your PR, but it should be a couple lines to adapt to the change in interface. Please rebase on top of that commit before submitting further changes.

Also, with regards to the rebase, you may want to wait for the commit to check for imports first. It should be up shortly.

At any rate, thanks for the work!

**Comment by goatpig** [comment]
Imports commit is up

**Comment by 151henry151** [comment]
I'm working on addressing each of these. Happy to be involved, Armory is legendary and I'm honored to have the opportunity to help fix it!

**Comment by 151henry151** [comment]
@goatpig ready for review

**Comment by goatpig** [comment]
tACK, will fix the test before reviewing the follow up PR (#770)
